Tmedia 676 input component

JackHowa opened the pull request:
## Ticket

- [TMEDIA-676](https://arcpublishing.atlassian.net/browse/TMEDIA-676)

## Description

Add Input component

## Acceptance Criteria

Create an <Input> component

Supported types: email, password, and text

Input includes label, input area, placeholder text, help text, and value

Includes state variations of default, warning, success and error 

Have Scss file that adds the hooks to the Sass library

Have test coverage

Have storybook documentation




## Test Steps

1. Checkout branch - `git checkout TMEDIA-676-input-component`
2. Update dependencies - `npm i`
3. Run Storybook `npm run storybook`
4. Check out the Input storybook documentation
5. Look at the example mocks to validate that the variations use make sense visually of what's possible. 

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed relevant documentation has been updated/added.
- [x] Add label - **ready for review** when the pull request is ready for someone to begin reviewing

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] All GitHub Actions pass
- [ ] Ran the code locally based on the test instructions.
- [ ] Checked Chromatic for Storybook changes, accepted the updates if acceptable
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.
- [ ] Approve and Add label - **ready to merge** if you are happy with the pull request
- [ ] Want another reviewer? Add the label **additional review**

JackHowa committed:
feat: Add starter for input component

JackHowa committed:
feat: Add basic functionality for input and label and types

JackHowa committed:
feat: Add bem naming for handling state variations nd elements within the sass library

JackHowa committed:
feat: Add style hooks for input states

JackHowa committed:
Merge branch 'arc-themes-release-version-2.00' into TMEDIA-676-input-component

JackHowa committed:
Merge branch 'arc-themes-release-version-2.00' into TMEDIA-676-input-component

JackHowa committed:
feat: Add placeholder style. Fix typo

JackHowa committed:
fix: Use correctmixin

JackHowa committed:
fix: Remove children prop

JackHowa committed:
docs: Remove example that's not visually displaying anything

JackHowa committed:
docs: Remove comment

matthewroach commented:
The blue design is the focus styles, so there are 4 variations, and a focus style

matthewroach commented on src/components/input/_index.scss:
```diff
@@ -0,0 +1,45 @@
+@use "../../scss";
+
```
I don't see any hooks for providing the ability to add styles for the label.

This doesn't quite seem quite right - as the comment is called `input` but we also need to expose the ability to style the `<input />` too

```css
.c-input {
	@include scss.component-properties("input");
}
```

JackHowa replied:
what? the inputs are styled via: 

```scss
.c-input__input--error {
	@include scss.component-properties("input-error");
}

.c-input__input--success {
	@include scss.component-properties("input-success");
}

.c-input__input--warning {
	@include scss.component-properties("input-warning");
}

// added placeholder for default status
// https://developer.mozilla.org/en-US/docs/Web/CSS/::placeholder
.c-input__input--default {
	@include scss.component-with-states("input-default", [placeholder]);
}

```

JackHowa replied:
will add a way to style the label -- good call. with this way, like the for the input or no? 

if not, should I use the actual elements as selectors `label {` rather than classes? Would need to use data properties then to get variations, I'd imagine

```jsx
`${COMPONENT_CLASS_NAME}__label--${derivedInputState}
```

JackHowa replied:
addressed and updated

matthewroach commented on src/components/input/index.jsx:
```diff
@@ -0,0 +1,159 @@
+	autoComplete,
+	className,
+	defaultValue = "",
+	hidden = false,
```
We should be consistent and use `defaultProps` instead of setting them inline

```
Input.defaultProps = { }
```

JackHowa replied:
ok will do

JackHowa replied:
addressed

JackHowa committed:
feat: Clarify input style hooks. Add label style hooks with variations

JackHowa committed:
feat: Use hidden for container class name. Add variations to base classes along with focus state of input.

Based on feedback https://github.com/WPMedia/arc-themes-components/pull/48#discussion_r828037013

JackHowa committed:
docs: Add comment for illustrating tip area reference

JackHowa committed:
fix: Fix typo from mixin component-with-states

JackHowa committed:
refactor: Use default props rather than fallback param values

Via https://github.com/WPMedia/arc-themes-components/pull/48#discussion_r828041225

JackHowa committed:
refactor: Nest classes within parent so that they can only be used together

JackHowa committed:
feat: Allow each variation to style its placeholder as well based on mocks

matthewroach commented on src/components/input/index.jsx:
```diff
@@ -0,0 +1,171 @@
+	return (
+		<div className={containerClassNames}>
+			<label
+				className={`${COMPONENT_CLASS_NAME}__label ${COMPONENT_CLASS_NAME}__label--${derivedInputState}`}
```
I am curious if we approach this is a different way - and apply the `${derivedInputState}` class to the parent element rather than each child and use the cascade to it's full advantage. That way if you have 

```
.c-input--error {
  color: red;
}
```

which is then applied to all children elements

JackHowa replied:
oh ok. this assumes that your label color would want to be used the same as your tip or error color, right?

matthewroach replied:
Right - and we could still add hooks for the label - just thinking through the way an editor would have to configure the least amount of settings.
Setting red color at the parent then having to only change the label if they want that to be different would be less configuration - but also remember we still want the individual control of each item to provide the most flexibility

JackHowa replied:
so we want both the parent variations and the children ones?

matthewroach replied:
Not exactly - something like this could work 

```css
@use "../../scss";

.c-input {
	@include scss.component-properties("input");

	.c-input__input {
		@include scss.component-properties("input-input");
	}

	.c-input__label {
		@include scss.component-properties("input-label");
	}

	.c-input__tip {
		@include scss.component-properties("input-tip");
	}
}

.c-input--error {
	@include scss.component-properties("input-error");

	.c-input__input {
		@include scss.component-properties("input-error-input");
	}

	.c-input__label {
		@include scss.component-properties("input-error-label");
	}

	.c-input__tip {
		@include scss.component-properties("input-error-tip");
	}
}
```

Note - the above is just a snippet of what the output would be, we would need to account for each state

JackHowa replied:
huh. do you think I should go down this for this starting ticket? To achieve the mocks, the implementation that just was pushed (modifying all the children) gets there

JackHowa replied:
huh. for achieving the mocks, I guess the parent styling all of the colors wouldn't work though. The label does not become red, for instance, on error. should I iterate through these based on your example? 

although, the label could always redefine its color at a more specific level that it's always black...

JackHowa replied:
updated @matthewroach

JackHowa committed:
feat: Refactor to use the top-level parent rather than the children to style individually

Based on https://github.com/WPMedia/arc-themes-components/pull/48#discussion_r828448632

JackHowa committed:
refactor: Allow individual styling for each variation from the parent level

Per https://github.com/WPMedia/arc-themes-components/pull/48#discussion_r828459508

JackHowa committed:
test: Add test coverage rewritten in rtl

JackHowa committed:
docs: Add example and intro text for the input

matthewroach commented on src/components/input/_index.scss:
```diff
@@ -0,0 +1,81 @@
+	}
+}
+
+.c-input:focus {
```
`:focus` is a state only applicable to `<input />` in this component we should update the Scss to have it applied to the input - similar to this

```scss
	.c-input__input {
		@include scss.component-with-states("input-warning-input", [focus]);
	}
```

JackHowa replied:
updated @matthewroach

JackHowa committed:
feat: Add focus only for input element. Add default variation as well in addition to the base formatting

Based on https://github.com/WPMedia/arc-themes-components/pull/48#discussion_r828979928

JackHowa committed:
fix: Use correct mixin spelling
